File: app/src/main/java/com/teixeira/subtitles/activities/MainActivity.java
package com.teixeira.subtitles.activities;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import androidx.fragment.app.Fragment;
import com.teixeira.subtitles.R;
import com.teixeira.subtitles.callbacks.UpdateProjectsCallback;
import com.teixeira.subtitles.databinding.ActivityMainBinding;
import com.teixeira.subtitles.fragments.ProjectsFragment;
import com.teixeira.subtitles.fragments.sheets.CreateProjectSheetFragment;

public class MainActivity extends BaseActivity implements UpdateProjectsCallback {

  private ActivityMainBinding binding;

  @Override
  protected View bindView() {
    binding = ActivityMainBinding.inflate(getLayoutInflater());
    return binding.getRoot();
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    setSupportActionBar(binding.toolbar);

    binding.fabNewProject.setOnClickListener(
        v -> CreateProjectSheetFragment.newInstance().show(getSupportFragmentManager(), null));
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    getMenuInflater().inflate(R.menu.activity_main_menu, menu);
    return super.onCreateOptionsMenu(menu);
  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {

    if (item.getItemId() == R.id.menu_about) {
      startActivity(new Intent(this, AboutActivity.class));
    }

    return super.onOptionsItemSelected(item);
  }

  @Override
  public void updateProjects() {
    Fragment fragment = getSupportFragmentManager().findFragmentById(binding.projects.getId());

    if (fragment != null) {
      ((ProjectsFragment) fragment).loadProjects();
    }
  }
}

File: app/src/main/java/com/teixeira/subtitles/activities/MainActivity.kt
package com.teixeira.subtitles.activities

import android.os.Bundle
import android.view.View
import androidx.activity.OnBackPressedCallback
import androidx.activity.viewModels
import com.google.android.material.color.MaterialColors
import com.teixeira.subtitles.R
import com.teixeira.subtitles.databinding.ActivityMainBinding
import com.teixeira.subtitles.fragments.dialogs.ProjectEditorDialogFragment
import com.teixeira.subtitles.handlers.PermissionsHandler
import com.teixeira.subtitles.viewmodels.MainViewModel

class MainActivity : BaseActivity() {

  private var _binding: ActivityMainBinding? = null
  private val binding: ActivityMainBinding
    get() = checkNotNull(_binding) { "MainActivity has been destroyed!" }

  private val mainViewModel by viewModels<MainViewModel>()

  private val onBackPressedCallback =
    object : OnBackPressedCallback(false) {
      override fun handleOnBackPressed() {
        if (mainViewModel.currentFragmentIndex != MainViewModel.FRAGMENT_PROJECTS_INDEX) {
          mainViewModel.currentFragmentIndex = MainViewModel.FRAGMENT_PROJECTS_INDEX
        }
      }
    }

  override val statusBarColor: Int
    get() =
      MaterialColors.getColor(this, com.google.android.material.R.attr.colorOnSurfaceInverse, 0)

  override val navigationBarColor: Int
    get() =
      MaterialColors.getColor(this, com.google.android.material.R.attr.colorOnSurfaceInverse, 0)

  override val navigationBarDividerColor: Int
    get() =
      MaterialColors.getColor(this, com.google.android.material.R.attr.colorOnSurfaceInverse, 0)

  override fun bindView(): View {
    _binding = ActivityMainBinding.inflate(layoutInflater)
    return binding.root
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setSupportActionBar(binding.toolbar)

    lifecycle.addObserver(PermissionsHandler(this, activityResultRegistry))
    onBackPressedDispatcher.addCallback(onBackPressedCallback)
    mainViewModel.observeCurrentFragmentIndex(this) { onFragmentChanged(it) }

    configureListeners()
  }

  override fun onDestroy() {
    super.onDestroy()
    onBackPressedCallback.isEnabled = false
    _binding = null
  }

  private fun configureListeners() {
    binding.toolbar.setNavigationOnClickListener { onBackPressedDispatcher.onBackPressed() }

    binding.fabNewProject.setOnClickListener {
      mainViewModel.currentFragmentIndex = MainViewModel.FRAGMENT_PROJECTS_INDEX
      ProjectEditorDialogFragment.newInstance().show(supportFragmentManager, null)
    }

    binding.bottomNavigation.setOnItemSelectedListener { item ->
      when (item.itemId) {
        R.id.item_projects ->
          mainViewModel.currentFragmentIndex = MainViewModel.FRAGMENT_PROJECTS_INDEX
        R.id.item_settings ->
          mainViewModel.currentFragmentIndex = MainViewModel.FRAGMENT_SETTINGS_INDEX
        else -> false
      }
      true
    }
  }

  private fun onFragmentChanged(fragmentIndex: Int) {
    val isProjectsFragment = fragmentIndex == MainViewModel.FRAGMENT_PROJECTS_INDEX
    if (isProjectsFragment) {
      supportActionBar?.setTitle(R.string.projects)
      setNavigationSelectedItem(R.id.item_projects)
    } else {
      supportActionBar?.setTitle(R.string.settings)
      setNavigationSelectedItem(R.id.item_settings)
    }
    binding.fragmentsContainer.displayedChild = fragmentIndex

    supportActionBar?.setDisplayHomeAsUpEnabled(!isProjectsFragment)
    supportActionBar?.setHomeButtonEnabled(!isProjectsFragment)
    onBackPressedCallback.isEnabled = !isProjectsFragment
  }

  private fun setNavigationSelectedItem(id: Int) {
    if (binding.bottomNavigation.selectedItemId != id) {
      binding.bottomNavigation.setSelectedItemId(id)
    }
  }
}
